dvc/utils/stream.py

```diff
@@ -43,3 +46,15 @@ def read1(self, n=-1):
 
         output, self.leftover = chunk[:n], chunk[n:]
         return output
+
+    read = read1
+
+    def peek(self, n: int):
+        data = b""
+        while len(data) < n:
+            chunk = self.read1(n - len(data))
+            if not chunk:
+                break
+            data += chunk
+        self.leftover = data + self.leftover if self.leftover else data
+        return data
```

Comment:
> Looks weird. If `len(self.leftover) >= n` we should simply return it otherwise we should grow it. Adding new data on the left doesn't make sense.

Whole point of peek is to return data, but don't move cursor, meaning that same data should be returned on next .read
Having that we should preserv whatever we return by peek in self.leftover
So, If `len(self.leftover) >= n` than read will pick it up, cut the extra and return as chunk, see no problems there, and tests shows the same.

Also peek should not return data less than request unless EOF is reached, that is why it is done in the loop there.

Reply:
There is no point to read anything if we have enough leftover. And you are not supposed to add old leftover after new data, this shuffles data. Tests simply don't cover.

Until you cut `self.leftover` you are not advancing position and it can be repeeked or reread.

Reply:
> There is no point to read anything if we have enough leftover. And you are not supposed to add old leftover after new data, this shuffles data. Tests simply don't cover.

It is covered right there:

>     with open_url((http / "sample.txt").url, mode="rb") as fd:
        text = text.encode("utf8")
        empty = b""
        assert fd.peek(len(text)) == text
        assert read_nbytes(fd, len(text), mode="rb") == text
        **assert fd.peek(len(text)) == text
        assert read_nbytes(fd, mode="rb") == text
        assert fd.read() == empty**

> Until you cut `self.leftover` you are not advancing position and it can be repeeked or reread.

Gotcha, added.